iceoryx_hoofs/include/iceoryx_hoofs/internal/log/platform_building_blocks/logger.inl

```diff
@@ -0,0 +1,137 @@
+// Copyright (c) 2019 by Robert Bosch GmbH. All rights reserved.
+// Copyright (c) 2021 - 2022 by Apex.AI Inc. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// SPDX-License-Identifier: Apache-2.0
+
+#ifndef IOX_HOOFS_PLATFORM_BUILDING_BLOCKS_LOGGER_INL
+#define IOX_HOOFS_PLATFORM_BUILDING_BLOCKS_LOGGER_INL
+
+#include "iceoryx_hoofs/log/platform_building_blocks/logger.hpp"
+
+#include <atomic>
+#include <cstdint>
+#include <cstring>
+#include <mutex>
+
+namespace iox
+{
+namespace pbb
+{
+template <uint32_t N>
+// NOLINTJUSTIFICATION see at declaration in header
+// NOLINTNEXTLINE(hicpp-avoid-c-arrays, cppcoreguidelines-avoid-c-arrays)
+inline constexpr bool equalStrings(const char* lhs, const char (&rhs)[N]) noexcept
+{
+    return strncmp(lhs, rhs, N) == 0;
+}
+
+
+template <typename BaseLogger>
+inline constexpr LogLevel Logger<BaseLogger>::minimalLogLevel() noexcept
+{
+    return Logger::MINIMAL_LOG_LEVEL;
+}
+
+template <typename BaseLogger>
+inline constexpr bool Logger<BaseLogger>::ignoreLogLevel() noexcept
+{
+    return Logger::IGNORE_ACTIVE_LOG_LEVEL;
+}
+
+
+template <typename BaseLogger>
+inline Logger<BaseLogger>& Logger<BaseLogger>::get() noexcept
+{
+    // NOLINTJUSTIFICATION needed for the functionality
+    // NOLINTNEXTLINE(cppcoreguidelines-avoid-non-const-global-variables)
+    thread_local static Logger* logger = Logger::activeLogger();
+    if (!logger->m_isActive.load(std::memory_order_relaxed))
+    {
+        // no need to loop until m_isActive is true since this is an inherent race
+        //   - the logger needs to be active for the whole lifetime of the application anyway
+        //   - if the logger was changed again, the next call will update the logger
+        //   - furthermore, it is not recommended to change the logger more than once
+        logger = Logger::activeLogger();
```

Comment:
OK, we detect that the logger is outdated, and hence change it for the active thread. Since both loggers are valid (i.e. exist in memory) it is fine to detect this late (which can happen).

It is ensured that eventually the logger will be updated, which is good enough (and performance-wise the best solution we can come up with, to my knowledge).